play-validations/memory-footprint/src/main/java/com/google/wear/watchface/dfx/memory/ArscTable.java

```diff
@@ -0,0 +1,129 @@
+import com.google.devrel.gmscore.tools.apk.arsc.TypeChunk;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Set;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Represents all the resources of interest in the Android package.
+ *
+ * These resources are loaded from the resources.arsc file. Where obfuscation has been applied, the
+ * mapping is derived, for the creation of the ArscResource objects. For example, for a logical
+ * resource res/raw/watchface.xml, the package may in fact store this as res/aB.xml. The
+ * resources.arsc file contains this mapping, and the ArscTable provides a list of resources with
+ * their logical types, names and data.
+ *
+ * Note that more than one ArscResource object can exist for the given dimensions. For example, if
+ * there is a drawable and a drawable-fr folder, then there may be multiple ArscResource entries for
+ * drawables with the same type, name and extension. The configuration detail, e.g. "fr" or
+ * "default", is not currently exposed in the ArscResource objects as it isn't used.
+ */
+public class ArscTable {
+    // Only certain resource types are of interest to the evaluator, notably, not string resources.
+    private static final Set<String> RESOURCE_TYPES = Set.of("raw", "xml", "drawable", "font", "asset");
+    private static final String RESOURCES_FILE_NAME = "resources.arsc";
+    private final List<ArscResource> resourceTable;
+
+    private ArscTable(List<ArscResource> resources) {
+        this.resourceTable = resources;
+    }
+
+    /**
+     * Creates the table from a path to an AAB structure on the file system.
+     * @param aabPath The path to the root of the AAB directory.
+     * @return The constructed table.
+     * @throws IOException when the resources file cannot be found, or other IO errors occur.
+     */
+    static ArscTable createFromAabDirectory(Path aabPath) throws IOException {
+        Path arscPath = Paths.get(aabPath.toString(), RESOURCES_FILE_NAME);
+        File arscFile = new File(arscPath.toAbsolutePath().toString());
+        if (!arscFile.exists()) {
+            throw new FileNotFoundException("Resources file not found");
+        }
+        InputStream is = new FileInputStream(arscFile);
+        return createTable(is, null, aabPath);
+    }
+
+    /**
+     * Creates the table from a AAB or APK file.
+     * @param zipFile The zip file object representing the AAB/APK.
+     * @return The constructed table.
+     * @throws IOException when the resources file cannot be found, or other IO errors occur.
+     */
+    static ArscTable createFromAndroidPackage(ZipFile zipFile) throws IOException {
+        ZipEntry arscEntry = new ZipEntry(RESOURCES_FILE_NAME);
+        InputStream is = zipFile.getInputStream(arscEntry);
```

Comment:
nit: probably doesn't actually matter, but try with resources to close the stream is a good idea.  Especially from a Zip file.

Reply:
Done, thanks!